### ISSUE TITLE
Importing photos from CSV files sent by Streetspotr

### ISSUE THREAD

**Issue by Svenyo** [sozialhelden/wheelmap#630]
So, here are the files with photos from Streetspotr from the current campaign.

There is one file for each country: Belgium (BE), Germany (DE), Denmark (DK), Spain (ES), France (FR), Italy (IT), Ireland (IR), Netherland (NL), United Kingdom (UK).

Please import them into the database.
[Streetspotr CSV files 062017.zip](https://github.com/sozialhelden/wheelmap/files/1077788/Streetspotr.CSV.files.062017.zip)

**Comment by Svenyo** [comment]
@1000miles @holgerd @anngreenberg If this is too short in notice for the current milestone list please consider this issue to be important for the next milestones starting on June, 19.

**Comment by 1000miles** [comment]
Thank you for the files, @Svenyo. We will test your new CSV files today and try to deploy them to staging. As soon as this is done, we will inform you to review them on staging. If everything works well, we can deploy them to production on monday (2017, june 19) after final approval.

**Comment by 1000miles** [comment]
Copied comment from #167:

@Svenyo cc @holgerd 

> IMPORTANT: We need to delete the broken images first before we can do new imports from csv via the rake task, otherwise the photo duplicate checker will not import the photos because it thinks the photo already exists (even though it is broken). We will wait for your check and final approval to remove the corrupt 14733 effected images before we import new csv data from streetspotr.

**Comment by 1000miles** [comment]
@Svenyo The files inside the zip file `Streetspotr CSV files 062017.zip` are all named `BE_2016.csv`, DE_2016.csv`, etc. Can you please confirm if `2016` data is correct?

**Comment by 1000miles** [comment]
### Update, 2017, 19 June:

Following data exist on staging at the moment:
- pois: 1.004.426
- provided_pois: 3.214
- photos: 25.819

**General rule: We can only import data from csv if the poi already exists in the database.**

Even though these data listed above on staging seem to be a lot it looks as if we do not have enough data on staging to do the test import. I did a dry-run  with `DE_2016.csv` that contains 19197 records. 

Here is an example output:

```bash
RAILS_ENV=staging bundle exec rake streetspotr:check file="DE_2016.csv"

POIS not found: 0.
POIS and OSM_ID nil: 19196.
CREATED: Photos: 0
CREATED: Provided Pois: 0
EXISTING BEFORE IMPORT: ProvidedPois: 3214, Photos: 25819.
EXISTING AFTER IMPORT: ProvidedPois: 3214, Photos: 25819.
```

**Comment by 1000miles** [comment]
Sorry, my bad. I thought I already renamed the column `attrId` to `osm_id`, so here is the dry-run result for the file `DE_2016.csv` (it would create 9333 Photos and 9744 Provided Pois):

```bash
$ RAILS_ENV=staging bundle exec rake streetspotr:check file="DE_2016.csv"

POIS not found: 4095.
POIS and OSM_ID nil: 5168.
CREATED: Photos: 9933
CREATED: Provided Pois: 9744
EXISTING BEFORE IMPORT: ProvidedPois: 3214, Photos: 25819.
```

**Comment by Svenyo** [comment]
@1000miles The data is correct, Streetspotr named the files probably "2016" because they made the setup of the campaign already in 2016.

**Comment by holgerd** [comment]
@1000miles 
> it would create 9333 Photos

In the CSV file @Svenyo and I counted 19.197 photo URIs (column named _Store photos (URL)_). What about the other photos? I think we have ca 2 photos per place.

**Comment by 1000miles** [comment]
### UPDATE, 2017-06-27

## Production

- [x] Import `BE_2016.csv` - status: done!

  ```
  POIS not found: 0.
  POIS and OSM_ID nil: 0.
  CREATED: Photos: 10
  CREATED: Provided Pois: 5
  ```
- [x] Import `DK_2016.csv` - status: done!
  
  ```
  POIS not found: 5.
  POIS and OSM_ID nil: 5.
  CREATED: Photos: 143
  CREATED: Provided Pois: 60
  ```

- [x] Import `ES_2016.csv` - status: done!

  ```
  POIS not found: 0.
  POIS and OSM_ID nil: 0.
  CREATED: Photos: 529
  CREATED: Provided Pois: 251
  ```

- [x] Import `FR_2016.csv` - status: done!

  ```
  POIS not found: 0.
  POIS and OSM_ID nil: 0.
  CREATED: Photos: 145
  CREATED: Provided Pois: 61
   ```
- [x] Import `IT_2016.csv` - status: done!

  ```
  POIS not found: 12.
  POIS and OSM_ID nil: 13.
  CREATED: Photos: 1431
  CREATED: Provided Pois: 630
  ```

- [x] Import `NL_2016.csv` - status: done!

  ```
  POIS not found: 6.
  POIS and OSM_ID nil: 8.
  CREATED: Photos: 182
  CREATED: Provided Pois: 84
  ```

- [x] Import `DE_2016.csv` - status: open
---

#IMPORT CURRENTLY BROKEN!

- Import `IR_2016.csv`
- Import `UK_2016.csv`

#### Error output for IR & UK:

```shell
rake aborted!
ArgumentError: argument out of range
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/values/time_zone.rb:292:in `initialize'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/values/time_zone.rb:292:in `new'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/values/time_zone.rb:292:in `parse'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/core_ext/string/zones.rb:9:in `in_time_zone'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/attribute_methods/time_zone_conversion.rb:64:in `convert_value_to_time_zone'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/attribute_methods/time_zone_conversion.rb:37:in `taken_at='
/var/apps/wheelmap/production/releases/20170626102954/app/models/photo.rb:53:in `extract_date_time'
...
```

**Comment by 1000miles** [comment]
> In the CSV file @Svenyo and I counted 19.197 photo URIs (column named Store photos (URL)). What about the other photos? I think we have ca 2 photos per place.

@holgerd that were stats from staging, let's see what production will say after did the big import of 19.197 records for `DE_2016.csv` today. :) 

cc @Svenyo @anngreenberg

**Comment by Svenyo** [comment]
@1000miles @holgerd @anngreenberg Ok, thanks for the stats so far. Let's see what the number of photos on production for Germany will be. :-D

**Comment by 1000miles** [comment]
Thanks, @Svenyo! :)

Running the import `DE_2016.csv` on production now, debugging `IR_2016.csv` and `UK_2016.csv` meanwhile. More infos later here.

**Comment by 1000miles** [comment]
### Update: 2017-06-28

Unfortunately the import for `DE_2016.csv` can not be finished as well because of the same error (`ArgumentError: argument out of range`) that points to the method  [`extract_date_time`](https://github.com/sozialhelden/wheelmap/blob/master/app/models/photo.rb#L52-L58).

## Production: 

**Observations so far:**

- `DE_2016.csv` stops the csv import after line **8995** with osm_id: **3312074516**

  ```ruby
  8995:3088240;;Ja-Su Kitchen - Art: Essen & Trinken;Steps-Spot (DE);Stufen-Spots DE   2016;2017-01-21 11:28:16 UTC;2017-01-21 15:12:12 UTC;2017-01-21 15:30:12   UTC;0.0;false;GroÃŸe BrunnenstraÃŸe 81, 22763 Hamburg;GroÃŸe BrunnenstraÃŸe   81;22763;Hamburg;Germany;324989;32;m;keine;Hamburg;HTC HTC 10;Android(24);1;Yes;https://download.streetspotr.com/1ac76747b3dc161639650a4107fbcb52c4fae052/944dd3973b6bae2a7ca159b7fb348c2362d87544;;;;2017-01-21 12:28:32 UTC;Yes;Yes;Yes;No;No;Nostalgia Bar;Yes;;;food;Essen & Trinken;Hamburg;Germany;3312074516;53.5526069;9.9237516;Ja-Su Kitchen;22763;GroÃŸe BrunnenstraÃŸe 81;restaurant;unknown
  ```

- `IR_2016.csv` stops the csv import before it even starts.

- `UK_2016.csv` stops the csv import  after line **161** with osm_id: **-82096288**

  ```ruby
  161:3441453;;BHS - Type: Shopping;Steps-Spot (UK);Stufen-Spots UK 2016;2017-05-06 05:39:34 UTC;2017-05-06 05:44:30 UTC;2017-05-06 06:00:05 UTC;0.0;false;30 Union Street, B2 4RG Birmingham;30 Union Street;B2 4RG;Birmingham;United Kingdom;529527;50;f;Bachelors degree;Birmingham;iPhone 5;iOS 10.3.2;1;No;https://download.streetspotr.com/1bd10908a626ace54fc943194e33f9b742c2185a/cfad455afbbd146d08ac9f226367144febbbef55;Store no longer open. Went bankrupt;52.479575;-1.8957027777777777;2017-05-06 06:40:31 UTC;;;;;;;;;Store no longer open to public. It has gone into bankruptcy;Shopping;shopping;United Kingdom;-82096288;52.479496;-1.8956103;BHS;department_store;unknown;Birmingham;B2 4RG;30 Union Street
  ```

- in `dev` AND `staging` mode the rake task loops successfully through the csv but this might be due to the fact that production has way more pois than `dev` and `staging`.

**Stats from today's import for `DE_2016.csv`, `IR_2016.csv` and `UK_2016.csv` before 	interruptions:**

```shell
CREATED: Photos: 2900
CREATED: Provided Pois: 2969
```

---

#### Full Error backtrace:

```shell
Skipped: POI for osm_id 3312074516 not found.
Skipped: osm_id blank and POI not found!
rake aborted!
ArgumentError: argument out of range
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/values/time_zone.rb:292:in `initialize'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/values/time_zone.rb:292:in `new'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/values/time_zone.rb:292:in `parse'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/core_ext/string/zones.rb:9:in `in_time_zone'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/attribute_methods/time_zone_conversion.rb:64:in `convert_value_to_time_zone'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/attribute_methods/time_zone_conversion.rb:37:in `taken_at='
/var/apps/wheelmap/production/releases/20170626102954/app/models/photo.rb:53:in `extract_date_time'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:429:in `block in make_lambda'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:161:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:161:in `block in halting'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:501:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:501:in `block in call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:501:in `each'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:501:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:86:in `run_callbacks'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/callbacks.rb:306:in `_create_record'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/timestamp.rb:57:in `_create_record'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/persistence.rb:484:in `create_or_update'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/callbacks.rb:302:in `block in create_or_update'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:113:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:113:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:552:in `block (2 levels) in compile'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:502:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:502:in `call'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activesupport-4.1.15/lib/active_support/callbacks.rb:86:in `run_callbacks'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/callbacks.rb:302:in `create_or_update'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/persistence.rb:125:in `save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/validations.rb:57:in `save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/attribute_methods/dirty.rb:29:in `save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/transactions.rb:273:in `block in save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/transactions.rb:329:in `block in with_transaction_returning_status'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `block in transaction'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/connection_adapters/abstract/database_statements.rb:209:in `within_new_transaction'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `transaction'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/transactions.rb:208:in `transaction'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/transactions.rb:326:in `with_transaction_returning_status'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/transactions.rb:273:in `save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/newrelic_rpm-3.15.0.314/lib/new_relic/agent/instrumentation/active_record_helper.rb:34:in `block in save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/newrelic_rpm-3.15.0.314/lib/new_relic/agent.rb:574:in `with_database_metric_name'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/newrelic_rpm-3.15.0.314/lib/new_relic/agent/instrumentation/active_record_helper.rb:33:in `save!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/protected_attributes-1.1.3/lib/active_record/mass_assignment_security/validations.rb:17:in `create!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/relation.rb:147:in `block in create!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/relation.rb:292:in `scoping'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/relation.rb:147:in `create!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/protected_attributes-1.1.3/lib/active_record/mass_assignment_security/relation.rb:61:in `find_or_create_by!'
/var/apps/wheelmap/production/shared/bundle/ruby/2.2.0/gems/activerecord-4.1.15/lib/active_record/querying.rb:6:in `find_or_create_by!'
/var/apps/wheelmap/production/releases/20170626102954/lib/tasks/streetspotr.rake:55:in `block (3 levels) in <top (required)>'
/var/apps/wheelmap/production/releases/20170626102954/lib/tasks/streetspotr.rake:30:in `block (2 levels) in <top (required)>'
Tasks: TOP => streetspotr:import
(See full trace by running task with --trace)
```

NEXT:
- [ ] investigate method `extract_date_time` and why the csv import stops at some points.

**Comment by 1000miles** [comment]
As of now following items were created by 2017-06-28:

```
CREATED: Photos: 8706 (incl. successful image processing)
CREATED: Provided Pois: 2969
```

**Comment by holgerd** [comment]
@1000miles what is the status of this? Can this be done as part of https://github.com/sozialhelden/wheelmap/milestone/26 as well? cc @anngreenberg

**Comment by 1000miles** [comment]
@holgerd cc @Svenyo @anngreenberg 

This ticket is on hold because it is dependent on https://github.com/sozialhelden/wheelmap/issues/633. As soon as #633 is done, we can continue with the import of the remaining files `IR_2016.csv`, `UK_2016.csv` and `DE_2016.csv` from this ticket. 

As agreed with @anngreenberg we are starting with following ticket first (starting from today): https://github.com/sozialhelden/wheelmap/issues/587

**Comment by 1000miles** [comment]
### Update, Mon, 2017-07-10

@Svenyo cc @holgerd @anngreenberg 

- [x] Import `IR_2016.csv` - status: done!

```shell
POIS not found: 1.
POIS and OSM_ID nil: 1.
CREATED: Photos: 70
CREATED: Provided Pois: 34
```

- [x] Import `UK_2016.csv` - status: done!

```shell
POIS not found: 14.
POIS and OSM_ID nil: 18.
CREATED: Photos: 1117
CREATED: Provided Pois: 482
```

- [x] Import `DE_2016.csv` - status: done!

```shell
POIS not found: 200.
POIS and OSM_ID nil: 242.
CREATED: Photos: 10052
CREATED: Provided Pois: 4265
```

We imported **11.239** photos today of which **7.807** photos currently still needs to go through the `image_processing` delayed_job task of the carrierwave backgrounder to create different image size versions. As soon as this is done we will leave a note here. 

@Svenyo Following format is currently not supported by the DateTime parser. Maybe interesting to know for streetspotr?

```shell
EXIF date_time format '2017:01:21 3:49:36 PM' is not supported: invalid date
EXIF date_time format '2017:01:21 3:49:36 +2000' is not supported: invalid date
```

**Comment by 1000miles** [comment]
### Update, 2017-07-11

@Svenyo cc @holgerd @anngreenberg 

The image_processing has been done for yesterday's created 11.239 photos.

Items created between `2017-06-27` and `2017-07-10`:
```
37.392 photos
11.258 provided_pois
```

**Comment by holgerd** [comment]
awesome!

**Comment by anngreenberg** [comment]
Imports worked as expected, therefore closing this issue.
